### PR TITLE
dont use mutables as default values, DRY POSIX IPC

### PR DESCRIPTION
**Pull request by yedpodtrzitko**
Hi,
the main motivation for creating this PR is a number of places where a mutable variable is used as a default value for a function params. I believe it's a widely known thing, but I'll put here an example anyway:
``` 
>>> def append_to(element, to=[]):
...     to.append(element)
...     return to
>>> print append_to(1)
[1]
>>> print append_to(2) # a sane output would be [1] 
[1, 2]
```
^ if this behaviour is utilized on purpose, then... I'm not sure if that's a good design, as it doesnt make the code easy to follow.

other things done here:
- some unused pieces of code are removed
- added missing `super()` call where it would make sense
- DRY duplicite code in POSIX IPC

I've "tested" the result code to build SM on OSX, and Linux

Regards,
yed_

### CODE REVIEW

**Comment by dvander** [comment]
Oh... oh god. Is that really Python behavior? That's beyond broken. Thanks for catching that.

Unrelated, it's hard to review stuff when there's a lot of miscellaneous changes. Big changesets also make it harder to bisect problems, since there are more things to diff through. In the future individual commits for each fix are a lot better.

File: ambuild2/frontend/amb2/gen.py [hunk 1]
```diff
@@ -317,7 +317,7 @@ def validateOutputFolder(self, path):
     raise Exception('illegal path component')
 
   def parseOutput(self, cwd_entry, path, kind):
-    if path[-1] == os.sep or path[-1] == os.altsep or path == '.' or path == '':
+    if path in ('.', '') or path[-1] in (os.sep, os.altsep):
```
Comment (dvander) [inline ambuild2/frontend/amb2/gen.py]: This change doesn't look necessary.

File: ambuild2/frontend/base/gen.py [hunk 2]
```diff
@@ -199,16 +199,15 @@ def compileScript(self, path):
       chars = fp.read()
 
       # Python 2.6 can't compile() with Windows line endings?!?!!?
-      chars = chars.replace('\r\n', '\n')
-      chars = chars.replace('\r', '\n')
+      chars = chars.replace('\r\n', '\n').replace('\r', '\n')
```
Comment (dvander) [inline ambuild2/frontend/base/gen.py]: The previous style is intentional - the sequence of operations is clearer before.

File: ambuild2/frontend/cpp/compilers.py [hunk 3]
```diff
@@ -59,7 +59,7 @@ def __init__(self, options = None):
   def inherit(self, other):
     self.debuginfo = other.debuginfo
     for attr in self.attrs:
-      setattr(self, attr, copy.copy(getattr(other, attr)))
+      setattr(self, attr, copy.deepcopy(getattr(other, attr)))
```
Comment (dvander) [inline ambuild2/frontend/cpp/compilers.py]: Good catch, though this is a big enough change to definitely warrant its own commit.

File: ambuild2/ipc/generic_poll.py [hunk 4]
```diff
@@ -17,23 +17,17 @@
...
-import multiprocessing as mp
+import select
 from . import process
 from . import posix_proc
-from . process import ProcessHost, Channel, Error, Special
+from . process import Error, Special
 
 # Generic POSIX multiprocess support, implemented using poll().
 
-class MessagePump(process.MessagePump):
```
Comment (dvander) [inline ambuild2/ipc/generic_poll.py]: This refactoring is significant, it should be its own PR.

File: ambuild2/util.py [hunk 5]
```diff
@@ -92,11 +92,12 @@ def WaitForProcess(process):
...
-  pargs = { 'args': argv }
-  pargs['stdout'] = subprocess.PIPE
-  pargs['stderr'] = subprocess.PIPE
-  if executable != None:
-    pargs['executable'] = executable
+  pargs = {
+    'args': argv,
+    'stdout': subprocess.PIPE,
+    'stderr': subprocess.PIPE,
+    'executable': executable,
```
Comment (dvander) [inline ambuild2/util.py]: This is technically a functional change since 'executable' is now a key where it wasn't before. I doubt it matters here though.

File: ambuild2/task.py [hunk 6]
```diff
@@ -66,12 +64,6 @@ def __init__(self, pump, channel, vars):
     })
 
   def receiveTask(self, channel, message):
-    task_id = message['task_id']
-    task_type = message['task_type']
-    task_folder = message['task_folder']
-    if not task_folder:
-      task_folder = '.'
```
Comment (dvander) [inline ambuild2/task.py]: What happened to this check here?

Why did the task_type variable get folded into an expression? (I generally split things like this out for easier debugging.)

File: ambuild2/util.py (hunk 5 above)
Comment (yedpodtrzitko) [inline ambuild2/util.py]: I am aware this isnt the exact thing as it was before, but in this case it doesnt matter, because the default value for `executable` is `None`. So the result will be the same.

File: ambuild2/task.py (hunk 6 above)
Comment (yedpodtrzitko) [inline ambuild2/task.py]: this variable is not used anywhere. Even if it would be, there's the same check a few lines below (`if message['task_folder']:`), so there's no need to make the same check twice, is it?

**Comment by yedpodtrzitko** [comment]
I agree the PR is fugly, so I'll split it into a few smaller PRs.
I'll be back
